src/lavinmq/queue/stream_queue_message_store.cr

```diff
@@ -88,6 +98,92 @@ module LavinMQ
+        positions = Hash(String, Int64).new
+        return positions if @consumer_offsets.size.zero?
+
+        loop do
+          ctag_length = @consumer_offsets.read_byte
+          break if !ctag_length || ctag_length.zero?
+          ctag = @consumer_offsets.read_string(ctag_length)
+          positions[ctag] = @consumer_offsets.pos
+          @consumer_offsets.skip(8)
+        rescue IO::EOFError
+          break
+        end
+        @consumer_offsets.pos = 0 if @consumer_offsets.pos == 1
+        @consumer_offsets.resize(@consumer_offsets.pos)
+        positions
+      end
+
+      def update_consumer_offset(consumer_tag : String, new_offset : Int64)
+        if pos = @consumer_offset_positions[consumer_tag]?
+          IO::ByteFormat::SystemEndian.encode(new_offset, @consumer_offsets.to_slice(pos, 8, false))
+        else
+          store_consumer_offset(consumer_tag, new_offset)
+        end
+      end
+
+      def store_consumer_offset(consumer_tag : String, new_offset : Int64)
+        slice = consumer_tag.to_slice
+        consumer_tag_length = slice.size.to_u8
+        pos = @consumer_offsets.size + slice.size + 1
+
+        length_buffer = uninitialized UInt8[1]
+        IO::ByteFormat::LittleEndian.encode(consumer_tag_length, length_buffer.to_slice)
+
+        offset_buffer = uninitialized UInt8[8]
+        IO::ByteFormat::LittleEndian.encode(new_offset.as(Int64), offset_buffer.to_slice)
+
+        @consumer_offsets.write(length_buffer.to_slice + slice + offset_buffer.to_slice)
+        @consumer_offset_positions[consumer_tag] = pos
+      end
+
+      def cleanup_consumer_offsets
+        offsets_to_save = Hash(String, Int64).new
+        lowest_offset_in_stream, _seg, _pos = offset_at(@segments.first_key, 4u32)
+        @consumer_offset_positions.each do |ctag, _pos|
+          offset = last_offset_by_consumer_tag(ctag)
+          next if !offset || offset < lowest_offset_in_stream
+          offsets_to_save[ctag] = offset
+        end
+
+        delete_and_reopen_offsets_file
+        @consumer_offset_positions = Hash(String, Int64).new
+        offsets_to_save.each do |ctag, offset|
+          store_consumer_offset(ctag, offset)
+        end
+      end
+
+      def remove_consumer_tag_from_file(consumer_tag : String)
+        @consumer_offset_positions = @consumer_offset_positions.reject! { |k, _v| k == consumer_tag }
+
+        offsets_to_save = Hash(String, Int64).new
+        @consumer_offset_positions.each do |ctag, _p|
+          offset = last_offset_by_consumer_tag(ctag)
+          next unless offset
+          offsets_to_save[ctag] = offset
+        end
+
+        delete_and_reopen_offsets_file
+        offsets_to_save.each do |ctag, offset|
+          store_consumer_offset(ctag, offset)
+        end
+      end
+
+      def delete_and_reopen_offsets_file
+        @consumer_offsets.close
+        @consumer_offsets.delete
+        @consumer_offsets = MFile.new(@consumer_offset_path, @consumer_offset_capacity)
+      end
```

Comment:
```suggestion
      def replace_offsets_file(&)
        old_consumer_offsets = @consumer_offsets
        @consumer_offsets = MFile.new("#{@consumer_offset_path}.tmp", @consumer_offset_capacity)
        yield # fill the new file with correct data in this block
        @consumer_offsets.rename @consumer_offset_path # this will delete the old file on disk
        old_consumer_offsets.close
      end
```